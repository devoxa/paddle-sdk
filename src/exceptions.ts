export class PaddleSdkException extends Error {}
