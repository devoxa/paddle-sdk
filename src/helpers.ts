import { serialize as phpSerialize } from 'php-serialize'
import dayjs from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import utc from 'dayjs/plugin/utc'
import FormData from 'form-data'

dayjs.extend(customParseFormat)
dayjs.extend(utc)

export function stableSerialize(object: { [key: string]: any }): string {
  // 1) Sort the object alphabetically by it's keys
  object = sortByKey(object)

  // 2) Encode arrays in their string form: `[1, 2, 3]` -> `'1, 2, 3'`
  // 3) Encode any non-strings as their JSON stringified version: `3` -> `'3'`
  for (const property in object) {
    if (object.hasOwnProperty(property) && typeof object[property] !== 'string') {
      if (Array.isArray(object[property])) {
        object[property] = object[property].join(', ')
      } else {
        object[property] = JSON.stringify(object[property])
      }
    }
  }

  // 4) Serialize in the way that PHP would
  return phpSerialize(object)
}

function sortByKey(object: { [key: string]: any }) {
  const keys = Object.keys(object).sort()

  const sortedObject: { [key: string]: any } = {}
  for (const i in keys) {
    sortedObject[keys[i]] = object[keys[i]]
  }

  return sortedObject
}

export function parseUtcDate(dateString: string, type: 'DATE' | 'DATE_TIME' | 'EXPIRY_DATE'): Date {
  switch (type) {
    case 'DATE':
      return dayjs.utc(dateString, 'YYYY-MM-DD').toDate()
    case 'DATE_TIME':
      return dayjs.utc(dateString, 'YYYY-MM-DD HH:mm:ss').toDate()
    case 'EXPIRY_DATE':
      return dayjs.utc(dateString, 'MM/YYYY').toDate()
  }
}

export function objectToFormData(object: { [key: string]: any }): FormData {
  const formData = new FormData()

  Object.entries(object).map(([key, value]) => {
    formData.append(key, value.toString())
  })

  return formData
}
