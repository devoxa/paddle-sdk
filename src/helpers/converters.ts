import {
  RawPaddleSubscriptionCreatedAlert,
  RawPaddleSubscriptionUpdatedAlert,
  RawPaddleSubscriptionPaymentSucceededAlert,
} from '../__generated__/webhook-alert-interfaces'
import {
  PaddleSdkSubscriptionStatus,
  PaddleSdkPausedReason,
  PaddleSdkPaymentMethod,
  PaddleSdkCurrency,
  PaddleSdkCountry,
} from '../interfaces'
import dayjs from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import utc from 'dayjs/plugin/utc'

dayjs.extend(customParseFormat)
dayjs.extend(utc)

export function convertApiInteger(integerString: string): number {
  return parseInt(integerString, 10)
}

export function convertApiFloat(floatString: string): number {
  return parseFloat(floatString)
}

export function convertApiBoolean(booleanString: '0' | '1'): boolean {
  return booleanString === '1'
}

export function convertApiDate(
  dateString: string,
  type: 'DATE' | 'DATE_TIME' | 'EXPIRY_DATE'
): Date {
  switch (type) {
    case 'DATE':
      return dayjs.utc(dateString, 'YYYY-MM-DD').toDate()
    case 'DATE_TIME':
      return dayjs.utc(dateString, 'YYYY-MM-DD HH:mm:ss').toDate()
    case 'EXPIRY_DATE':
      return dayjs.utc(dateString, 'MM/YYYY').toDate()
  }
}

export function convertApiSubscriptionStatus(
  subscriptionStatus: RawPaddleSubscriptionCreatedAlert['status']
): PaddleSdkSubscriptionStatus {
  switch (subscriptionStatus) {
    case 'active':
      return 'ACTIVE'
    case 'trialing':
      return 'TRIALING'
    case 'past_due':
      return 'PAST_DUE'
    case 'paused':
      return 'PAUSED'
    case 'deleted':
      return 'CANCELLED'
  }
}

export function convertSdkSubscriptionStatus(
  subscriptionStatus: PaddleSdkSubscriptionStatus
): RawPaddleSubscriptionCreatedAlert['status'] {
  switch (subscriptionStatus) {
    case 'ACTIVE':
      return 'active'
    case 'TRIALING':
      return 'trialing'
    case 'PAST_DUE':
      return 'past_due'
    case 'PAUSED':
      return 'paused'
    case 'CANCELLED':
      return 'deleted'
  }
}

export function convertApiPausedReason(
  pausedReason: RawPaddleSubscriptionUpdatedAlert['paused_reason']
): PaddleSdkPausedReason {
  switch (pausedReason) {
    case 'delinquent':
      return 'DELINQUENT'
    case 'voluntary':
      return 'VOLUNTARY'
  }
}

export function convertApiCurrency(currency: string): PaddleSdkCurrency {
  // These are the currencies already returned by paddle, we just make them type-safe
  return currency as PaddleSdkCurrency
}

export function convertApiCountry(country: string): PaddleSdkCountry {
  // These are the countries already returned by paddle, we just make them type-safe
  return country as PaddleSdkCountry
}

export function convertApiPaymentMethod(
  paymentMethod: RawPaddleSubscriptionPaymentSucceededAlert['payment_method']
): PaddleSdkPaymentMethod {
  switch (paymentMethod) {
    case 'card':
      return 'CARD'
    case 'paypal':
      return 'PAYPAL'
    case 'apple-pay':
      return 'APPLE_PAY'
    case 'wire-transfer':
      return 'WIRE_TRANSFER'
    case 'free':
      return 'FREE'
  }
}

export function convertApiCardType(cardType: string): string {
  // TODO Add exact types here
  return cardType.toUpperCase()
}
