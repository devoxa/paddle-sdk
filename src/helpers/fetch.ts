import nodeFetch from 'node-fetch'
import FormData from 'form-data'

type Body = Record<string, string | number | undefined>

export async function fetch<T>(url: string, options: { method: string; body: Body }): Promise<T> {
  const response = await nodeFetch(url, {
    method: options.method,
    body: objectToFormData(options.body),
  })

  return response.json()
}

function objectToFormData(object: Body): FormData {
  const formData = new FormData()

  Object.entries(object).forEach(([key, value]) => {
    if (typeof value === 'undefined') return

    formData.append(key, value.toString())
  })

  return formData
}
