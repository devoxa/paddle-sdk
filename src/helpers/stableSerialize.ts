import { serialize as phpSerialize } from 'php-serialize'

export function stableSerialize<T>(object: Record<string, T>): string {
  // 1) Sort the object alphabetically by it's keys
  object = sortByKey(object)

  // 2) Encode arrays in their string form: `[1, 2, 3]` -> `'1, 2, 3'`
  // 3) Encode any non-strings as their JSON stringified version: `3` -> `'3'`
  const encodedObject: Record<string, string> = {}
  for (const property in object) {
    if (object.hasOwnProperty(property) && typeof object[property] !== 'string') {
      const value = object[property]

      if (Array.isArray(value)) {
        encodedObject[property] = value.join(', ')
      } else {
        encodedObject[property] = JSON.stringify(value)
      }
    }
  }

  // 4) Serialize in the way that PHP would
  return phpSerialize(object)
}

function sortByKey<T>(object: Record<string, T>) {
  const keys = Object.keys(object).sort()

  const sortedObject: Record<string, T> = {}
  for (const i in keys) {
    sortedObject[keys[i]] = object[keys[i]]
  }

  return sortedObject
}
