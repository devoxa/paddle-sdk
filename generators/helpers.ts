import fetch from 'node-fetch'

type NextData = {
  props: {
    pageProps: {
      dehydratedState: {
        queries: Array<NextDehydratedStateQuery>
      }
    }
  }
}

type NextDehydratedStateQuery = {
  queryHash: string
  state: {
    data: {
      data: string
    }
  }
}

export async function fetchPageMarkdown(url: string) {
  const response = await fetch(url)
  const text = await response.text()

  const scriptMatch = text.match(/<script id="__NEXT_DATA__" [^>]*>(.*?)<\/script>/)
  if (!scriptMatch) throw new Error('Could not find __NEXT_DATA__ script')

  const data = JSON.parse(scriptMatch[1]) as NextData
  const queries = data.props.pageProps.dehydratedState.queries

  const urlQueryKey = url.split('/').pop() as string
  const queryHashRegex = new RegExp(`"[a-z0-9]{13}-${urlQueryKey}`)
  const query = queries.find((query) => query.queryHash.match(queryHashRegex))
  if (!query) throw new Error(`Could not find query for "${urlQueryKey}" in dehydratedState`)

  return query.state.data.data
}

export function spliceAndReturn<T>(array: Array<T>, index: number): T {
  const element = array[index]
  array.splice(index, 1)
  return element
}

export function trim(string: string, character: string) {
  let start = 0
  let end = string.length

  while (start < end && string[start] === character) ++start
  while (end > start && string[end - 1] === character) --end

  return start > 0 || end < string.length ? string.slice(start, end) : string
}

export function toPascalCase(string: string) {
  const camelCase = string.replace(/_(\w)/g, (_, $1) => $1.toUpperCase())
  return `${camelCase.charAt(0).toUpperCase()}${camelCase.slice(1)}`
}

export function toJSDoc(lines: Array<string>) {
  return lines.length ? `/**\n${lines.map((line) => `* ${line}`).join('\n')}\n*/` : ''
}
