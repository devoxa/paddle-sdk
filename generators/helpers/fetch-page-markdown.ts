import fetch from 'node-fetch'

type NextData = {
  props: {
    pageProps: {
      dehydratedState: {
        queries: Array<NextDehydratedStateQuery>
      }
    }
  }
}

type NextDehydratedStateQuery = {
  queryHash: string
  state: {
    data: {
      data: string
    }
  }
}

export async function fetchPageMarkdown(url: string) {
  const response = await fetch(url)
  const text = await response.text()

  const scriptMatch = text.match(/<script id="__NEXT_DATA__" [^>]*>(.*?)<\/script>/)
  if (!scriptMatch) throw new Error('Could not find __NEXT_DATA__ script')

  const data = JSON.parse(scriptMatch[1]) as NextData
  const queries = data.props.pageProps.dehydratedState.queries

  const urlQueryKey = url.split('/').pop() as string
  const query = queries.find((query) => query.queryHash.includes(urlQueryKey))
  if (!query) throw new Error(`Could not find query for "${urlQueryKey}" in dehydratedState`)

  return query.state.data.data
}
