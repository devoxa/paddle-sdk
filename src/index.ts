import { createVerify } from 'crypto'
import { stableSerialize } from './helpers/stableSerialize'
import {
  RawPaddleWebhookAlert,
  RawPaddleSubscriptionCreatedAlert,
  RawPaddleSubscriptionUpdatedAlert,
  RawPaddleSubscriptionCancelledAlert,
  RawPaddleSubscriptionPaymentSucceededAlert,
} from './__generated__/webhook-alert-interfaces'
import { PaddleSdkException, PaddleSdkApiException } from './exceptions'
import {
  PaddleSdkSubscriptionCreatedAlert,
  PaddleSdkSubscriptionUpdatedAlert,
  PaddleSdkSubscriptionCancelledAlert,
  PaddleSdkSubscriptionPaymentSucceededAlert,
  PaddleSdkCreateProductPayLinkRequest,
  PaddleSdkListSubscriptionsRequest,
  PaddleSdkUpdateSubscriptionRequest,
  PaddleSdkCancelSubscriptionRequest,
  PaddleSdkCreateSubscriptionModifierRequest,
  PaddleSdkCreateProductPayLinkResponse,
  PaddleSdkListSubscriptionsResponse,
  PaddleSdkUpdateSubscriptionResponse,
  PaddleSdkCancelSubscriptionResponse,
  PaddleSdkCreateSubscriptionModifierResponse,
} from './interfaces'
import {
  RawPaddlePostProductGeneratePayLinkRequest,
  RawPaddlePostProductGeneratePayLinkResponse,
  PADDLE_PRODUCT_GENERATE_PAY_LINK,
  RawPaddlePostSubscriptionUsersUpdateRequest,
  RawPaddlePostSubscriptionUsersUpdateResponse,
  PADDLE_SUBSCRIPTION_USERS_UPDATE,
  RawPaddlePostSubscriptionModifiersCreateRequest,
  RawPaddlePostSubscriptionModifiersCreateResponse,
  PADDLE_SUBSCRIPTION_MODIFIERS_CREATE,
  RawPaddlePostSubscriptionUsersCancelRequest,
  RawPaddlePostSubscriptionUsersCancelResponse,
  PADDLE_SUBSCRIPTION_USERS_CANCEL,
  RawPaddlePostSubscriptionUsersRequest,
  RawPaddlePostSubscriptionUsersResponse,
  PADDLE_SUBSCRIPTION_USERS,
} from './__generated__/api-route-interfaces'
import { encrypt, decrypt } from '@devoxa/aes-encryption'
import { fetch } from './helpers/fetch'
import {
  convertApiInteger,
  convertApiCurrency,
  convertApiDate,
  convertApiBoolean,
  convertApiSubscriptionStatus,
  convertApiFloat,
  convertApiPausedReason,
  convertApiPaymentMethod,
  convertApiCardType,
  convertSdkSubscriptionStatus,
} from './helpers/converters'

export * from './interfaces'

export interface PaddleSdkOptions {
  publicKey: string
  vendorId: number
  vendorAuthCode: string
  passthroughEncryptionKey: string
}

export class PaddleSdk<TPassthrough = any> {
  private readonly publicKey: string
  private readonly vendorId: number
  private readonly vendorAuthCode: string
  private readonly passthroughEncryptionKey: string

  constructor(options: PaddleSdkOptions) {
    if (!options.publicKey) {
      throw new PaddleSdkException('PaddleSdk was called without a publicKey')
    }

    if (!options.vendorId) {
      throw new PaddleSdkException('PaddleSdk was called without a vendorId')
    }

    if (!options.vendorAuthCode) {
      throw new PaddleSdkException('PaddleSdk was called without a vendorAuthCode')
    }

    if (!options.passthroughEncryptionKey || options.passthroughEncryptionKey.length !== 32) {
      throw new PaddleSdkException('PaddleSdk was called with an invalid passthroughEncryptionKey')
    }

    this.publicKey = options.publicKey
    this.vendorId = options.vendorId
    this.vendorAuthCode = options.vendorAuthCode
    this.passthroughEncryptionKey = options.passthroughEncryptionKey
  }

  verifyWebhookAlert(body: any): body is RawPaddleWebhookAlert {
    if (typeof body !== 'object') return false

    const { p_signature: signature, ...postBodyRest } = body || {}
    if (!signature || typeof signature !== 'string') return false

    const serializedPostBody = stableSerialize(postBodyRest)

    const verifier = createVerify('sha1')
    verifier.update(serializedPostBody)
    verifier.end()

    return verifier.verify(this.publicKey, signature, 'base64')
  }

  parseWebhookAlert(body: any) {
    if (!this.verifyWebhookAlert(body)) {
      throw new PaddleSdkException('Failed validating alert body')
    }

    switch (body.alert_name) {
      case 'subscription_created':
        return this.parseSubscriptionCreatedWebhookAlert(body)
      case 'subscription_updated':
        return this.parseSubscriptionUpdatedWebhookAlert(body)
      case 'subscription_cancelled':
        return this.parseSubscriptionCancelledWebhookAlert(body)
      case 'subscription_payment_succeeded':
        return this.parseSubscriptionPaymentSucceededWebhookAlert(body)
    }

    throw new PaddleSdkException(
      `Implementation missing: Can not parse alert of type ${body.alert_name}`
    )
  }

  private stringifyPassthrough(passthrough: any) {
    return encrypt(this.passthroughEncryptionKey, JSON.stringify(passthrough))
  }

  private parsePassthrough(passthrough: string): TPassthrough {
    try {
      return JSON.parse(decrypt(this.passthroughEncryptionKey, passthrough))
    } catch (err) {
      throw new PaddleSdkException('Failed parsing passthrough: ' + err.message)
    }
  }

  private parseSubscriptionCreatedWebhookAlert(
    body: RawPaddleSubscriptionCreatedAlert
  ): PaddleSdkSubscriptionCreatedAlert<TPassthrough> {
    return {
      alert_id: convertApiInteger(body.alert_id),
      alert_name: 'SUBSCRIPTION_CREATED',
      cancel_url: body.cancel_url,
      checkout_id: body.checkout_id,
      currency: convertApiCurrency(body.currency),
      email: body.email,
      event_time: convertApiDate(body.event_time, 'DATE_TIME'),
      marketing_consent: convertApiBoolean(body.marketing_consent),
      next_bill_date: convertApiDate(body.next_bill_date, 'DATE'),
      passthrough: this.parsePassthrough(body.passthrough),
      quantity: convertApiInteger(body.quantity),
      source: body.source,
      status: convertApiSubscriptionStatus(body.status),
      subscription_id: convertApiInteger(body.subscription_id),
      subscription_plan_id: convertApiInteger(body.subscription_plan_id),
      unit_price: convertApiFloat(body.unit_price),
      update_url: body.update_url,
      user_id: convertApiInteger(body.user_id),
    }
  }

  private parseSubscriptionUpdatedWebhookAlert(
    body: RawPaddleSubscriptionUpdatedAlert
  ): PaddleSdkSubscriptionUpdatedAlert<TPassthrough> {
    return {
      alert_id: convertApiInteger(body.alert_id),
      alert_name: 'SUBSCRIPTION_UPDATED',
      cancel_url: body.cancel_url,
      checkout_id: body.checkout_id,
      currency: convertApiCurrency(body.currency),
      email: body.email,
      event_time: convertApiDate(body.event_time, 'DATE_TIME'),
      marketing_consent: convertApiBoolean(body.marketing_consent),
      new_next_bill_date: convertApiDate(body.next_bill_date, 'DATE'),
      new_price: convertApiFloat(body.new_price),
      new_quantity: convertApiInteger(body.new_quantity),
      new_status: convertApiSubscriptionStatus(body.status),
      new_subscription_plan_id: convertApiInteger(body.subscription_plan_id),
      new_unit_price: convertApiFloat(body.new_unit_price),
      old_next_bill_date: convertApiDate(body.old_next_bill_date, 'DATE'),
      old_price: convertApiFloat(body.old_price),
      old_quantity: convertApiInteger(body.old_quantity),
      old_status: convertApiSubscriptionStatus(body.old_status),
      old_subscription_plan_id: convertApiInteger(body.old_subscription_plan_id),
      old_unit_price: convertApiFloat(body.old_unit_price),
      passthrough: this.parsePassthrough(body.passthrough),
      paused_at: body.paused_at ? convertApiDate(body.paused_at, 'DATE_TIME') : null,
      paused_from: body.paused_from ? convertApiDate(body.paused_from, 'DATE_TIME') : null,
      paused_reason: body.paused_reason ? convertApiPausedReason(body.paused_reason) : null,
      subscription_id: convertApiInteger(body.subscription_id),
      update_url: body.update_url,
      user_id: convertApiInteger(body.user_id),
    }
  }

  private parseSubscriptionCancelledWebhookAlert(
    body: RawPaddleSubscriptionCancelledAlert
  ): PaddleSdkSubscriptionCancelledAlert<TPassthrough> {
    return {
      alert_id: convertApiInteger(body.alert_id),
      alert_name: 'SUBSCRIPTION_CANCELLED',
      cancellation_effective_date: convertApiDate(body.cancellation_effective_date, 'DATE'),
      checkout_id: body.checkout_id,
      currency: convertApiCurrency(body.currency),
      email: body.email,
      event_time: convertApiDate(body.event_time, 'DATE_TIME'),
      marketing_consent: convertApiBoolean(body.marketing_consent),
      passthrough: this.parsePassthrough(body.passthrough),
      quantity: convertApiInteger(body.quantity),
      status: convertApiSubscriptionStatus(body.status),
      subscription_id: convertApiInteger(body.subscription_id),
      subscription_plan_id: convertApiInteger(body.subscription_plan_id),
      unit_price: convertApiFloat(body.unit_price),
      user_id: convertApiInteger(body.user_id),
    }
  }

  private parseSubscriptionPaymentSucceededWebhookAlert(
    body: RawPaddleSubscriptionPaymentSucceededAlert
  ): PaddleSdkSubscriptionPaymentSucceededAlert<TPassthrough> {
    return {
      alert_id: convertApiInteger(body.alert_id),
      alert_name: 'SUBSCRIPTION_PAYMENT_SUCCEEDED',
      balance_currency: convertApiCurrency(body.balance_currency),
      balance_earnings: convertApiFloat(body.balance_earnings),
      balance_fee: convertApiFloat(body.balance_fee),
      balance_gross: convertApiFloat(body.balance_gross),
      balance_tax: convertApiFloat(body.balance_tax),
      checkout_id: body.checkout_id,
      country: body.country,
      coupon: body.coupon,
      currency: convertApiCurrency(body.currency),
      customer_name: body.customer_name,
      earnings: convertApiFloat(body.earnings),
      email: body.email,
      event_time: convertApiDate(body.event_time, 'DATE_TIME'),
      fee: convertApiFloat(body.fee),
      initial_payment: convertApiBoolean(body.initial_payment),
      instalments: convertApiInteger(body.instalments),
      marketing_consent: convertApiBoolean(body.marketing_consent),
      next_bill_date: convertApiDate(body.next_bill_date, 'DATE'),
      next_payment_amount: convertApiFloat(body.next_payment_amount),
      order_id: body.order_id,
      passthrough: this.parsePassthrough(body.passthrough),
      payment_method: convertApiPaymentMethod(body.payment_method),
      payment_tax: convertApiFloat(body.payment_tax),
      plan_name: body.plan_name,
      quantity: convertApiInteger(body.quantity),
      receipt_url: body.receipt_url,
      sale_gross: convertApiFloat(body.sale_gross),
      status: convertApiSubscriptionStatus(body.status),
      subscription_id: convertApiInteger(body.subscription_id),
      subscription_payment_id: convertApiInteger(body.subscription_payment_id),
      subscription_plan_id: convertApiInteger(body.subscription_plan_id),
      unit_price: convertApiFloat(body.unit_price),
      user_id: convertApiInteger(body.user_id),
    }
  }

  private async apiRequest<TRequest, TResponse>(
    url: string,
    method: 'GET' | 'POST',
    body: TRequest
  ): Promise<TResponse> {
    const json = await fetch(url, {
      method,
      body: {
        vendor_id: this.vendorId,
        vendor_auth_code: this.vendorAuthCode,
        ...body,
      },
    })

    // Turn errors from the Paddle API into a unique exception with the error message
    if (json.success === false) {
      throw new PaddleSdkApiException(json.error.message)
    }

    return json.response
  }

  async createProductPayLink(
    data: PaddleSdkCreateProductPayLinkRequest<TPassthrough>
  ): Promise<PaddleSdkCreateProductPayLinkResponse> {
    return this.apiRequest<
      RawPaddlePostProductGeneratePayLinkRequest,
      RawPaddlePostProductGeneratePayLinkResponse
    >(PADDLE_PRODUCT_GENERATE_PAY_LINK.url, PADDLE_PRODUCT_GENERATE_PAY_LINK.method, {
      ...data,
      passthrough: this.stringifyPassthrough(data.passthrough),
    })
  }

  async listSubscriptions(
    data: PaddleSdkListSubscriptionsRequest
  ): Promise<PaddleSdkListSubscriptionsResponse> {
    function formatPaymentInformation(
      paymentInformation: RawPaddlePostSubscriptionUsersResponse[0]['payment_information']
    ) {
      if (paymentInformation.payment_method === 'card') {
        return {
          payment_method: 'CARD' as const,
          card_type: convertApiCardType(paymentInformation.card_type),
          last_four_digits: paymentInformation.last_four_digits,
          expiry_date: convertApiDate(paymentInformation.expiry_date, 'EXPIRY_DATE'),
        }
      }

      if (paymentInformation.payment_method === 'paypal') {
        return {
          payment_method: 'PAYPAL' as const,
        }
      }

      // @ts-expect-error TS error here because we handled all types that should exist according to the API docs
      const message = `Unknown payment method "${paymentInformation.payment_method}" for subscriptions`
      throw new PaddleSdkException(message)
    }

    function formatPayment(payment: RawPaddlePostSubscriptionUsersResponse[0]['last_payment']) {
      return {
        amount: payment.amount,
        currency: convertApiCurrency(payment.currency),
        date: convertApiDate(payment.date, 'DATE'),
      }
    }

    const formatSubscription = (subscription: RawPaddlePostSubscriptionUsersResponse[0]) => {
      return {
        subscription_id: subscription.subscription_id,
        plan_id: subscription.plan_id,
        user_id: subscription.user_id,
        user_email: subscription.user_email,
        marketing_consent: subscription.marketing_consent,
        status: convertApiSubscriptionStatus(subscription.state),
        signup_date: convertApiDate(subscription.signup_date, 'DATE_TIME'),
        update_url: subscription.update_url,
        cancel_url: subscription.cancel_url,
        paused_at: subscription.paused_at
          ? convertApiDate(subscription.paused_at, 'DATE_TIME')
          : null,
        paused_from: subscription.paused_from
          ? convertApiDate(subscription.paused_from, 'DATE_TIME')
          : null,
        payment_information: formatPaymentInformation(subscription.payment_information),
        last_payment: formatPayment(subscription.last_payment),
        next_payment: subscription.next_payment ? formatPayment(subscription.next_payment) : null,
      }
    }

    return this.apiRequest<
      RawPaddlePostSubscriptionUsersRequest,
      RawPaddlePostSubscriptionUsersResponse
    >(PADDLE_SUBSCRIPTION_USERS.url, PADDLE_SUBSCRIPTION_USERS.method, {
      subscription_id: data.subscription_id?.toString(),
      plan_id: data.plan_id?.toString(),
      state: data.status ? convertSdkSubscriptionStatus(data.status) : undefined,
      page: data.page,
      results_per_page: data.results_per_page,
    }).then((subscriptions) => subscriptions.map(formatSubscription))
  }

  async updateSubscription(
    data: PaddleSdkUpdateSubscriptionRequest<TPassthrough>
  ): Promise<PaddleSdkUpdateSubscriptionResponse> {
    return this.apiRequest<
      RawPaddlePostSubscriptionUsersUpdateRequest,
      RawPaddlePostSubscriptionUsersUpdateResponse
    >(PADDLE_SUBSCRIPTION_USERS_UPDATE.url, PADDLE_SUBSCRIPTION_USERS_UPDATE.method, {
      ...data,
      passthrough: this.stringifyPassthrough(data.passthrough),
    })
  }

  async cancelSubscription(
    data: PaddleSdkCancelSubscriptionRequest
  ): Promise<PaddleSdkCancelSubscriptionResponse> {
    return this.apiRequest<
      RawPaddlePostSubscriptionUsersCancelRequest,
      RawPaddlePostSubscriptionUsersCancelResponse
    >(PADDLE_SUBSCRIPTION_USERS_CANCEL.url, PADDLE_SUBSCRIPTION_USERS_CANCEL.method, data)
  }

  async createSubscriptionModifier(
    data: PaddleSdkCreateSubscriptionModifierRequest
  ): Promise<PaddleSdkCreateSubscriptionModifierResponse> {
    return this.apiRequest<
      RawPaddlePostSubscriptionModifiersCreateRequest,
      RawPaddlePostSubscriptionModifiersCreateResponse
    >(PADDLE_SUBSCRIPTION_MODIFIERS_CREATE.url, PADDLE_SUBSCRIPTION_MODIFIERS_CREATE.method, data)
  }
}
