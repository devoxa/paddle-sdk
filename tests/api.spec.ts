import * as FIXTURES from './fixtures'
import { PaddleSdk } from '../src/index'
import { fetch } from '../src/helpers/fetch'

jest.mock('../src/helpers/fetch', () => ({ fetch: jest.fn() }))

function mockNextApiResponse(response: any) {
  ;(fetch as jest.Mock).mockImplementationOnce(async () => ({ success: true, response }))
}

function getLastApiRequest() {
  return (fetch as jest.Mock).mock.calls[0]
}

describe('webhooks -> api methods', () => {
  const paddleSdk = new PaddleSdk({
    publicKey: FIXTURES.publicKey,
    vendorId: FIXTURES.vendorId,
    vendorAuthCode: FIXTURES.vendorAuthCode,
    passthroughEncryptionKey: FIXTURES.passthroughEncryptionKey,
  })

  beforeEach(() => {
    ;(fetch as jest.Mock).mockClear()
  })

  it('listSubscriptions', async () => {
    mockNextApiResponse(FIXTURES.listSubscriptionsApiResponse)

    const response = await paddleSdk.listSubscriptions({
      subscription_id: 123,
      plan_id: 123,
      status: 'CANCELLED',
      page: 0,
      results_per_page: 100,
    })

    expect(response).toMatchSnapshot()
    expect(getLastApiRequest()).toMatchSnapshot()
  })
})
