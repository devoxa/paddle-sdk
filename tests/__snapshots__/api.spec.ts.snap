// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webhooks -> api methods listSubscriptions 1`] = `
Array [
  Object {
    "cancel_url": "https://checkout.paddle.com/subscription/cancel?user=18124595&subscription=7908253&hash=9A4EE7D7645833AB6F90F81287DA92E139FE8DF6AF2090081357104138EC6937",
    "last_payment": Object {
      "amount": 2.99,
      "currency": "USD",
      "date": 2020-08-08T00:00:00.000Z,
    },
    "marketing_consent": false,
    "next_payment": null,
    "paused_at": null,
    "paused_from": null,
    "payment_information": Object {
      "card_type": "VISA",
      "expiry_date": 2021-08-01T00:00:00.000Z,
      "last_four_digits": "1824",
      "payment_method": "CARD",
    },
    "plan_id": 124890,
    "signup_date": 2020-08-08T23:15:48.000Z,
    "status": "CANCELLED",
    "subscription_id": 7908253,
    "update_url": "https://checkout.paddle.com/subscription/update?user=18124595&subscription=7908253&hash=AEA72A29FABC23692043D8B6D55FD29C625F532552AA8AC6EB137E604FC13E1A",
    "user_email": "foo@bar.com",
    "user_id": 18124595,
  },
]
`;

exports[`webhooks -> api methods listSubscriptions 2`] = `
Array [
  "https://vendors.paddle.com/api/2.0/subscription/users",
  Object {
    "body": Object {
      "page": 0,
      "plan_id": "123",
      "results_per_page": 100,
      "state": "deleted",
      "subscription_id": "123",
      "vendor_auth_code": "FooBarBaz",
      "vendor_id": 123456,
    },
    "method": "POST",
  },
]
`;
